syncer/syncer.go

```diff
@@ -1001,36 +1021,47 @@ func (s *Syncer) syncDDL(tctx *tcontext.Context, queueBucket string, db *DBConn,
 		if err != nil {
 			s.execError.Store(err)
 			if !utils.IsContextCanceledError(err) {
-				err = s.handleEventError(err, sqlJob.startLocation, sqlJob.currentLocation, true, sqlJob.originSQL)
+				err = s.handleEventError(err, ddlJob.startLocation, ddlJob.currentLocation, true, ddlJob.originSQL)
 				s.runFatalChan <- unit.NewProcessError(err)
 			}
 			s.jobWg.Done()
 			continue
 		}
 		s.jobWg.Done()
-		s.addCount(true, queueBucket, sqlJob.tp, int64(len(sqlJob.ddls)))
+		s.updateReplicationLag(ddlJob)
+		s.addCount(true, queueBucket, ddlJob.tp, int64(len(ddlJob.ddls)))
 	}
 }
 
-func (s *Syncer) sync(tctx *tcontext.Context, queueBucket string, db *DBConn, jobChan chan *job) {
+// DML synced in batch by one worker.
+func (s *Syncer) syncDML(tctx *tcontext.Context, queueBucket string, db *DBConn, jobChan chan *job) {
 	defer s.wg.Done()
 
 	idx := 0
 	count := s.cfg.Batch
 	jobs := make([]*job, 0, count)
 	tpCnt := make(map[opType]int64)
 
+	// clearF is used to calc lag metric and clear local job queue.
 	clearF := func() {
-		for i := 0; i < idx; i++ {
-			s.jobWg.Done()
+		if len(jobs) > 0 {
+			// NOTE: we can use the first job of job queue to calc lag because when this job flush to target db,
+			// every event before this job's event has already flushed to target db.
+			s.updateReplicationLag(jobs[0])
+		} else {
+			s.updateReplicationLag(nil)
 		}
-
-		idx = 0
-		jobs = jobs[0:0]
+		// calc tps
 		for tpName, v := range tpCnt {
 			s.addCount(true, queueBucket, tpName, v)
 			tpCnt[tpName] = 0
 		}
+		// reset job queue
+		for i := 0; i < idx; i++ {
+			s.jobWg.Done()
+		}
+		idx = 0
+		jobs = jobs[0:0]
```

Comment:
Why do we switch the position of these two parts?

Reply:
for clean logic, `for tpName, v := range tpCnt ` is to calc tps, the second part is to reset job queue